src/address.rs

```diff
@@ -210,99 +185,79 @@ impl<A, Rc: RefCounter> Address<A, Rc> {
 
     /// Waits until this address becomes disconnected.
     pub fn join(&self) -> ActorJoinHandle {
-        ActorJoinHandle(self.ref_counter.disconnect_notice())
+        ActorJoinHandle(self.0.disconnect_notice())
+    }
+
+    /// Returns true if this address and the other address point to the same actor. This is
+    /// distinct from the implementation of `PartialEq` as it ignores reference count type, which
+    /// must be the same for `PartialEq` to return `true`.
+    pub fn same_actor<Rc2: RefCounter>(&self, other: &Address<A, Rc2>) -> bool {
+        self.0.inner_ptr() == other.0.inner_ptr()
     }
 }
 
 /// A future which will complete when the corresponding actor stops and its address becomes
 /// disconnected.
 #[must_use = "Futures do nothing unless polled"]
-pub struct ActorJoinHandle(DropNotice);
+pub struct ActorJoinHandle(Option<EventListener>);
 
 impl Future for ActorJoinHandle {
     type Output = ();
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        self.0.poll_unpin(cx)
-    }
-}
-
-impl<A, M, Rc> Sink<M> for Address<A, Rc>
-where
-    A: Handler<M, Return = ()>,
-    M: Send + 'static,
-    Rc: RefCounter,
-{
-    type Error = Disconnected;
-
-    fn poll_ready(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        Pin::new(&mut self.sink)
-            .poll_ready(cx)
-            .map_err(|_| Disconnected)
-    }
-
-    fn start_send(mut self: Pin<&mut Self>, item: M) -> Result<(), Self::Error> {
-        let item = AddressMessage::Message(Box::new(NonReturningEnvelope::new(item)));
-        Pin::new(&mut self.sink)
-            .start_send(item)
-            .map_err(|_| Disconnected)
-    }
-
-    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        Pin::new(&mut self.sink)
-            .poll_flush(cx)
-            .map_err(|_| Disconnected)
-    }
-
-    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        Pin::new(&mut self.sink)
-            .poll_close(cx)
-            .map_err(|_| Disconnected)
+        match self.0.take() {
+            Some(mut listener) => match listener.poll_unpin(cx) {
+                Poll::Ready(()) => Poll::Ready(()),
+                Poll::Pending => {
+                    self.0 = Some(listener);
+                    Poll::Pending
+                }
+            },
+            None => Poll::Ready(()),
+        }
```

Comment:
To be honest, I didn't even look for one, so maybe I missed one. I can take a look later, unless you've already had a look?

Reply:
I looked and there isn't one :)

This is almost like `Fuse` but `Fuse` will emit `Pending` after completion.

Reply:
Ah, unfortunate! I guess that could be considered correct behaviour even for this, but if you are going to poll the future multiple times, I think `Ready` maybe makes more sense?

Reply:
I was about to say I agree with you but now that I think about it: Should we even guard the user here? It is a well-known thing to not poll futures again after they have emitted `Ready`. For cases where that cannot be guaranteed, there is `Fuse`.

Without trying it out and skimming the code of `EventListener`, it looks like it is designed to always allow at least one `poll`. Perhaps we can remove the entire `Option` dance and just store the `EventListener` inside?

Reply:
The `Option` is currently here to allow us to create a join handle that resolves immediately when the actor is already stopped (see tx.rs line 166). After the actor is stopped, the event listener will not get notified again, so creating with a `None` allows it to still resolve instead of pending forever. When implementing the future, though, I thought it was probably not worth adding a separate case to handle polling after Ready as Pending. That being said, maybe it's worth it

Reply:
Ah, thanks for the clarification.